### ISSUE TITLE
Notify the share owner when another user downloads a shared file [$1,000]

### ISSUE THREAD

**Issue by gig13** [owncloud/core#13158]
Use Case:
1. User A shares a file with User B.
2. User B downloads the file via Web/desktop client/mobile.
3. User A is notified that User B has downloaded the file and via which method - Web, desktop client or mobile.

It would also be desired to have this tracked in an audit trail like the Activity or via a reporting mechanism.

@MTRichards

**Comment by LukasReschke** [comment]
Ref https://github.com/owncloud/core/issues/7812 for public link sharing

**Comment by derekblankmccoy** [comment]
@LukasReschke Yea, for public link sharing, but more importantly for normal sharing.

This is an absolute must for a project I'm working on. Shared files with users must have traceability of who downloaded them and when. It's not so much to see if a user read an important document, but mainly to check who HASN'T read it. "Did you get the memo?" "No" "I shared it with the group" "Oh, well I didn't get it".
There's an app that can sort of do it called "file counter", but it doesn't work with oc8, and theres another one called "eslog" that sends hooks to elasticsearch but it doesn't record file reads as it doesn't seem to be a filesystem hook that gets recorded.

*** Willing to put a bounty on this feature ***

**Comment by derekblankmccoy** [comment]
Is there any development on this?

**Comment by bklo** [comment]
This feature is required on my project too. The FileCounter app does not suit what is needed. Any good news or should I look elsewhere?

**Comment by gig13** [comment]
@bklo 
This is not being developed. Closing out.

**Comment by gig13** [comment]
@bboule @MTRichards 
Please confirm if this should have a milestone.
Related to SF 2531

**Comment by bboule** [comment]
@MTRichards would need your guidance as to milestone?  I am assuming 9.0 is full?  @gig13 I am assuming this is a low severity as it seems cosmetic?

**Comment by MTRichards** [comment]
But this has traceability of who did what - the user agent string and the log file should track downloads...does it not do so?

I do not want to put this in the activity app and clutter that interface, but the log app for 9.0 should have the user agent string (the specific client and user that downloaded what). This should resolve this.

**Comment by bboule** [comment]
@MTRichards if I am reading this right, and I like to think I am, I think the ask here is if I share a file with you, I would like to be notified if you have downloaded the file and how (via my activity stream) that is to say I would like to be able to have notification (which I can turn on or off via personal page) of activity around items that I have shared with others... this seems reasonable I think what is being suggested is from the user perspective... clutter can be avoided by turning off in personal page... this would only impact the "shared"filter where I as a user can see who has accessed files I have shared....

Does this make sense?

**Comment by MTRichards** [comment]
Yep. I was keying off of this, which was only part of the comment:
>Shared files with users must have traceability of who downloaded them and when

We already have this for public link sharing, which was asked above, so this is covered in activity. 
And we have traceability in the log files for everything else, including user agent string in 9.0 logging.

If they want notification (inside activity), this is a different user story and it has to be added to the notifications options in the personal page as another checkbox. I don't want us to confuse traceability and compliance with notification, the activity app is too easy to break and mess with to be considered true traceability for audit and compliance purposes. As long as this is fully understood, and is turned off by default, we can look at adding this hook to the activity app.

9.0 is full and well underway, could consider for 9.1 but given how fill 9.1 is, it might be tough to squeeze it in organically. Might be something that could be squeezed in around the edges since it affects the activity app (from consulting), but again I think capacity constraints are the issue there too.

**Comment by derekblankmccoy** [comment]
To be clear. This is NOT ABOUT PUBLIC SHARING.
This is tractability for local sharing. If someone has an important memo that they have shared internally with their group, that person will want to know who has and hasn't read it (downloaded it).

I have used elasticsearch to log filesystem activity but downloads are not recorded. To some, this type of activity is more important than any other type.

**Comment by MTRichards** [comment]
Are you using community edition? In that case, this is correct. Enterprise logging should be downloading and recording all of this in the ownCloud log files.

**Comment by derekblankmccoy** [comment]
Oh I see, so download logging is an Enterprise edition only feature?

**Comment by MTRichards** [comment]
it is designed into the logs for enterprise edition, yes. The activity stream truncates itself, and a bunch of other things which can make it unreliable for logging and true traceability, thus my comments above.

**Comment by derekblankmccoy** [comment]
So can the hook be switched on for community edition so elasticsearch can see it? Not show anything on the on the UI's activity feed, but if someone bothers to set up elastic search, they can see whats happening. The same way Elasticsearch can see a whole bunch of other data that doesnt make it to the activity feed. It'll just be one extra bit.

**Comment by bboule** [comment]
@gig13 do you have what you need based on the disposition in SF tickets?  is this still needing information?

**Comment by gig13** [comment]
@bboule At this time I am all set, waiting for additional feedback from originating user

**Comment by PVince81** [comment]
The audit app should support this in 9.0, doesn't it ? @blizzz @nickvergessen 

Not sure about activities though.

**Comment by nickvergessen** [comment]
Well we have the full path in admin_audit now, so that should work.

For activities we don't have the full path and neither a way to resolve it to a share owner.

But since the issue says "activity or reporting app" I think it's okay and we can close it?

**Comment by derekblankmccoy** [comment]
@nickvergessen is this something that can be picked up by looking at non enterprise edition logs?

**Comment by blizzz** [comment]
@PVince81 well, it does not inform the owner, but the admin. If he looks into the log.

**Comment by derekblankmccoy** [comment]
Im thinking of putting a bounty on this.

Just as a recap:
ownCloud Community Edition, not Enterprise.
I share an important memo with my team of 10 people.
I want to be able to see who has/hasn't downloaded it.

As for interface, maybe a button can be added to the share drop down for that file, next to the trash can. pressing it can bring a popup box with everyone's name in the group and a timestamp for when it was downloaded. This would also have to work if the file was downloaded manually or automatically using the mobile app. 

I think this would require the filesystem "post_read" hook put back in (it was removed in OC8).

I'll wait for replies before putting the bounty on. I also have another OC user who is interested, so he might contribute. We might be looking at a 4 figure bounty here.

**Comment by derekblankmccoy** [comment]
#17495

**Comment by oparoz** [comment]
If I understand correctly, this is akin to the "signature folder" passing around in a corporate environment.

In terms of UI, I think it would be best to keep the info in a tab in the detailsview. You get a list of people who have been given access to the file and a check mark next to the ones who have downloaded it, along with the time stamp.

This could probably be a separate app, installable from the app store. This would give the project owner (derekblankmccoy and co) more freedom in terms of design and implementation.

**Comment by derekblankmccoy** [comment]
@oparoz You got it, exactly that. Need to know that a user has read a document. I understand that is may not be technically possible, but the fact that it's been downloaded is enough.
Like you say, in the details pane would be the perfect place to put it. 
Maybe colored arrows that match the android app:
Green = downloaded latest version
Yellow = downloaded but not latest

If file opened (viewed) is possible, that would be fantastic.

**Comment by oparoz** [comment]
@derekblankmccoy - Looking at the available hooks, I think we can track who has downloaded the file, but in order to be able to tell if it's been opened, then you'll need to booby-trap it. That might work with PDFs or Office files, but could be blocked by the user and is probably out of scope for this project. I don't think we have technology which can reliably tell if users have really read the document though ;).

I'm tempted to work on this if nobody else has started already, but one problem is that the sponsor's requirements are a bit different that the ones found in the OP (added notifications)

**Comment by derekblankmccoy** [comment]
@oparoz I think your idea of having it in the details pane makes the most sense. Keep i mind that when this issue was opened, the details pane did not exist. I mean, a notification tick could be added to the notification settings, then the owner can get an email. The issue is if 50 files are shared with 50 people, that's 250 emails going out

**Comment by oparoz** [comment]
@derekblankmccoy - I hear you, but this is a blue-ticket, opened by somebody else, so unless the solution is approved by the OP's client, nobody gets paid :| . One solution is to transfer the bounty to a separate issue in which we could clearly scope the project to match the sponsor's requirements. This ticket could then add notifications, if still required.
Another solution would be for @gig13 to agree with you :)

**Comment by derekblankmccoy** [comment]
Ok, I'll wait for @gig13 to reply. If we have to transfer it, I'll write up the requirements, ready to open as a new issue.

**Comment by jospoortvliet** [comment]
It would be enough to add this to the Activity stream of the file, right? Just like it already has the ability to warn you about the download of a publicly shared file, this issue asks to have it for ALL files. Well, if it's off by default, I think it makes for a nice little feature.

That seems to be exactly what @gig13 asked for and if @derekblankmccoy is OK with that solution and puts up the bounty, he gets to decide if the solution is satisfactory anyway, no matter who opened the original ticket. I'd say - go ahead, @oparoz ;-)

**Comment by derekblankmccoy** [comment]
That could work, but it needs to say who downloaded it with a timestamp. 
Also, looking at the activity tracker, how long can it get? Is there an easy way to export it to a csv?

At the moment, I'm using the esLog app and the closest I can get is seeing if a user logged in. I can only assume from there that they can see the shared file.
I don't even mind if the hooks get implemented so that elasticsearch can see it, but esLog no longer works with the latest OC.

**Comment by oparoz** [comment]
>That could work, but it needs to say who downloaded it with a timestamp. 

That would be part of the requirements, yes.

>Also, looking at the activity tracker, how long can it get?

This answer was given above:
"The activity stream truncates itself, and a bunch of other things which can make it unreliable for logging and true traceability"

but I haven't looked for that limit yet. Maybe it could be lifted if the user enables the feature, but there is also this statement (emphasis mine): 
"the activity app is **too easy to break and mess with** to be considered true traceability for audit and compliance purposes."

> Is there an easy way to export it to a csv?

I don't think a CSV exporter is available. That would require a script or occ command.

**Comment by derekblankmccoy** [comment]
Ok, how about your idea of a separate app?
I mean esLog would be perfect if it recorded file reads and if it worked with the latest version of OC.
Maybe an app that logs file activity and exports it to a csv daily and an interface that can visualise the data.

**Comment by oparoz** [comment]
>Ok, how about your idea of a separate app?

It gives you the flexibility to implement what you need, but there is something I forgot to mention. Beware that there is no guarantee that it will be maintained in a year from now unlike solutions implemented in `core`. Ultimately, you could end up in the same situation you are today with ESlog.
Unless of course if I and others start selling support for those add-ons and people actually subscribe to get updates.

>I mean esLog would be perfect if it recorded file reads and if it worked with the latest version of OC.

I took a look and there is too much "Illegal" code in that app, so it wouldn't be a good idea to patch it.

>Maybe an app that logs file activity and exports it to a csv daily and an interface that can visualise the data.

That would be a different app that the one adding a "signature panel", but interesting nonetheless.
Something like a table with a file on each row and users in columns, with colour coded cells?
If you end up with a large table, you'll want filtering and in the end it may not be better than using search in Files and clicking on the details view icon of the files in the results to view the stats.

**Comment by oparoz** [comment]
>I took a look and there is too much "Illegal" code in that app, so it wouldn't be a good idea to patch it.

Actually that may not be true. The app is outdated and should be rewritten using the AppFramework, but if what's making it incompatible with oC9.0 is outdated hooks, then fixing those could make the app generate the logs you're after.

**Comment by derekblankmccoy** [comment]
@oparoz There is another app called "file counter" that sort of did the job, but that one doesn't work with oC8 and above.

I know there is emphasis on keeping the interface clean, so it could be kept that way. Just put the hooks in and make it log that activity in a csv.

A really cool idea would be Slack integration, that way Slack can take care of the logging.

**Comment by oparoz** [comment]
@derekblankmccoy files_counter doesn't do what you need as it only counts files downloaded from the Files app. The results would be hugely inaccurate.

I could deliver an oC9 compatible app written with the Appframework which would log to ElasticSearch the following information when a shared file is opened (to download or sync):  current user, file, timestamp. I'm sure lots of ESlog features can be kept too, but I don't want this piece of work to be about trying to fix everything wrong in that app. I'll let you define what you need.
I would probably need to include activity notifications in order for @gig13 to mark this as solved unless we transfer the development of this app to a separate request.

There are alternatives, but you would need some solid processes in place and people who follow them.

.

**Comment by derekblankmccoy** [comment]
@oparoz 
I mentioned ESlog because it was almost there, to be honest ElasticSearch is a bit complicated and it logs a whole load of stuff that makes no sense. Here's the list of stuff it logged:

User login
User logout
User creation
User deletion
Password change
Group creation
Group deletion
User added/removed to/from a group
Read file **(it doesn't)**
Write file
Delete file
Copy/rename file
Share file
Enable/disablle 3rd party apps

But when Chekcing Kibana (the ElasticSearch visualizer), there's a bunch of other stuff that makes it a mess. You can eventually get to the info you're looking for, but it takes a lot of legwork. If file reads were actually there, I would have persevered with it and not put the bounty up.

Ideally, a Slack plugin for oC Monitoring would be fantastic. Is Slack integration for Activity something that you can do? The app could let you select what to log: file writes, file reads, fie deletes, file renames. We could look at possibly extending the feature to cover user logins at a future date if you wanted to do a separate piece of work (I could raise a separate issue and attach a bounty to it).

The OP mentions:
> have this tracked in an audit trail like the Activity or via a reporting mechanism.

Slack would be ideal, but an app that can display it in a table that can be sorted and filtered could also work. At the very least, have it log to a csv file.

**Comment by oparoz** [comment]
@derekblankmccoy - OK, as mentioned earlier, ES requests can be tuned to send exactly what you need, but if that complicates your workflow, then there is no reason to take that path.

>Ideally, a Slack plugin for oC Monitoring would be fantastic

AFAIK, that would be the same job as doing an app which connects to ES, I would just change the client. Someone could even extend the app to make it possible to pick the backend to use for logging.
Since you're paying, we could stick with Slack first, but IFTTT could maybe make it more useful (if technically possible) to others as it doesn't change much in terms of privacy (you're going through 3rd party providers to log your cloud activities), but allows admins to connect to the service they need.
You could also use the open source Slack alternative.

>Is Slack integration for Activity something that you can do?

The way I see it, that would be a different thing. You would batch process events instead of doing real-time monitoring.

>Slack would be ideal, but an app that can display it in a table that can be sorted and filtered could also work. At the very least, have it log to a csv file.

Yes, there are lots of possibilities :). I'll wait for a final Statement of Work against which my work can be validated before going further. I now have a decent understanding of what can be achieved.

**Comment by derekblankmccoy** [comment]
Ok, whichever way it goes, does oC core need to be modified at all to acknowledge filesystem reads? If so that has to be done whatever the outcome is.

**Comment by oparoz** [comment]
>Ok, whichever way it goes, does oC core need to be modified at all to acknowledge filesystem reads?

For reads of files shared, no. There could be a bug in core as others are reporting issues receiving some events in their apps, but that would be a bug which needs fixing in core, not something new which needs to be approved and implemented. There is a one line patch which can be applied anyway to work around the problem and which I used for testing.

**Comment by jospoortvliet** [comment]
One thought about this - now the Activity stream might have downsides (it isn't meant to be a legal-requirements-compliant monitoring tool), but it is enough for normal human beings and there IS a API to grab the content - the desktop client shows the content right now, for example and the mobile clients will introduce this over the coming releases. So you'll have access to this trace of 'what is going on' and 'who downloaded what' from multiple places. If it can be made to do what you need, maintenance is also guaranteed in the long run - a big plus over a one-off app. Just 2 cents ;-)

**Comment by derekblankmccoy** [comment]
@jospoortvliet The desktop client activity is no different to the activity feed on the web console. It doesn't show if a user I shared a document with has actually pressed the download button (on web or mobile client).
It's very simple. When I share something with someone, I want to know if they got it or not. It's the whole point of sharing after all.

Scenario:There's a group of 500 students and who are already using ownCloud for course work and sharing between themselves. They are participating in an obstacle race for charity. The organizer sends directions, instructions, safety documents, and application forms through ownCloud by doing an internal share to that group of students. 400 students apply successfully. 100 claim they never got the application form. The organizer should have a way of seeing that those people did in fact download it. Let's say that out of the 400 that applied, 50 don't turn up because they say there was no directions. The organizer can then check if those people downloaded the directions. Now let's say another 10 didn't wear the appropriate clothes because the instructions didn't mention anything about it. The organizer then easily finds out that they never downloaded the instructions in the first place.

This is the sort of information that is needed. Being able to see who has and who hasn't downloaded something that was shared with them, how they downloaded it (web or mobile), and a timestamp.

**Comment by oparoz** [comment]
@derekblankmccoy - I ran through a similar scenario when trying to understand if using the Activity panel was usable, but I thought it would quickly be unmanageable without a lot of customisation.
Let's say you get 48 notifications. How do you find the missing 2 if you need to send them a reminder?

But if I understand correctly what @jospoortvliet is saying. Pushing notifications to Activity, as he's suggesting, would make those available to all clients (web, desktop and mobile), making it easy for event organisers to find the information on their mobile devices, without having to use the web interface or an office document app. 
That's only useful if the problem I mentioned above can be fixed within the Activity app and ported to all clients of course.

**Comment by derekblankmccoy** [comment]
@oparoz So I guess the aim is to fix the activity app to include the owner's local share activity. From there, either the activity app would have to be fixed so it's got a reliable record, or we need external logging (elasticsearch, greylog, slack, whatever)

**Comment by oparoz** [comment]
@nickvergessen - What's missing in Activity to make it more robust and/or to make sure all records are kept?

**Comment by derekblankmccoy** [comment]
> The activity stream truncates itself, and a bunch of other things which can make it unreliable for logging and true traceability.

**Comment by oparoz** [comment]
Yes, but he's the author, so he may have some leads from a technical point of view.
Was the limit arbitrary in order to not fill the DB or was in put in place for performance reasons. Is it time based?, etc.

There is no point in trying to fix Activity if it will never be able to meet the requirements.

**Comment by derekblankmccoy** [comment]
Ok, so we're back to creating an app?

**Comment by oparoz** [comment]
Not necessarily. I'm hoping whatever limit is in place can easily be lifted using a setting (to be written). You will want the full log and will size your hardware accordingly if required, but a home user will probably be happy with the last 20 events or something.

The one thing I want to avoid is to log everything in Activity only to find out later that it doesn't work for you.

**Comment by oparoz** [comment]
### Custom app

* Can log to the DB (as good or bad as Activity)
* Can log to an external service (SLA determined by 3rd party)
* Could offer an API
* Best effort maintenance without subscription
* No need to get any sort of green light by the core team
* Should not require any hacking of core
* May not meet the OP's requirements

### Activity app

* Currently truncating logs
* Deemed unreliable by Product Manager
* Logs to the DB (may have scaling issues if adding more events?)
* Offers the information via an API
* API is used by official clients to get the list of events
* Officially maintained

**Comment by nickvergessen** [comment]
Well we have a background job that prunes after 365days by default. Value can be changed via config:
https://github.com/owncloud/activity/blob/master/lib/backgroundjob/expireactivities.php#L64-L66

The otherthing is downloading via webdav (syncing with client or federated share) does currently not trigger any hook, so they are not detectable.

**Comment by derekblankmccoy** [comment]
> The otherthing is downloading via webdav (syncing with client or federated share) does currently not trigger any hook, so they are not detectable.

This is the deal breaker. Unless that hook is there, this is dead.

**Comment by oparoz** [comment]
Thanks @nickvergessen. 365 days sound good enough for a feature which wants to track which files were downloaded.

> The otherthing is downloading via webdav (syncing with client or federated share) does currently not trigger any hook, so they are not detectable.

I saw the issue, but wasn't sure if that was because it wasn't implemented or if apps needed to register a different way. This should be relatively easy to fix since a short patch can give any app access to the information
@DeepDiver1975 @PVince81 @rullzer Is it on the 9.1 roadmap for dav (remote.php) to fire hooks?

**Comment by oparoz** [comment]
>This is the deal breaker. Unless that hook is there, this is dead.

@derekblankmccoy - It's a requirement for this work which is targeting 9.1, so the fix should be implemented in the coming weeks, but it's always possible to patch.

**Comment by oparoz** [comment]
@derekblankmccoy, @gig13 - I'm going to add an entry for each download of a shared file to the activity stream.

**Comment by oparoz** [comment]
@derekblankmccoy @gig13

### Global stream

![global_activities](https://cloud.githubusercontent.com/assets/323220/14109530/c0f18f78-f5c2-11e5-87b7-d7b6a3ccf9d0.png)

### File stream

![file_activities](https://cloud.githubusercontent.com/assets/323220/14109538/cb284892-f5c2-11e5-86a2-8283b3abaf7c.png)

### Desktop client stream

![desktop_activities](https://cloud.githubusercontent.com/assets/323220/14109545/d309137a-f5c2-11e5-81d9-16ef18e6ec21.png)

**Comment by oparoz** [comment]
@nickvergessen - Activity can listen to hooks triggered by DAV just fine. People with 3rd party apps were having issues registering for these events, but it turned out to be a documentation issue and does not require any `core` modifications.

**Comment by derekblankmccoy** [comment]
That's great! So it is possible. We still need to know how it was downloaded (web, client or mobile) and a way of exporting the info so we can have timestamps etc.
@gig13 is still active here, hopefully he will give some input

**Comment by oparoz** [comment]
>We still need to know how it was downloaded (web, client or mobile) 

![client_used](https://cloud.githubusercontent.com/assets/323220/14140336/ae7ec9b2-f678-11e5-9f1a-8a2168929351.png)

>and a way of exporting the info so we can have timestamps etc.

The client already exports that information via the "Copy" button when looking at the Server Activity, so I'm guessing you need something different?

**Comment by derekblankmccoy** [comment]
You are correct, I forgot about the client, as I don't use it. I only use web and mobile. I've just looked at the client and it shows a timestamp, the filename, the folder, the action and the size, but not who downloaded it. That would require changes to be made to the client (?)

An idea would be to add a filter in the web portal's activity feed that let's you select from last hour, day, week, month or year. And an equivalent of the desktop client's "copy" function. Maybe an "Export" button that dumps it in a csv? What do you think?

**Comment by oparoz** [comment]
>, but not who downloaded it. That would require changes to be made to the client (?)

The log extract I posted above is from the standard client and shows "who" and "using what", so no change required afaik.

>And an equivalent of the desktop client's "copy" function. Maybe an "Export" button that dumps it in a csv? What do you think?

You can retrieve all of it via any REST client and slice and dice it any way you want to, but maybe an ownCloud app could be written with the sole purpose of generating a daily CSV file.

Here is what it looks like for one item

```
{
	"activity_id": 1910,
	"type": "remote_download",
	"user": "NSA",
	"affecteduser": "admin",
	"app": "files_sharing",
	"subject": "shared_file_downloaded",
	"subjectparams": [{
		"value": {
			"56142": "\/Contracts\/Governments\/Most_wanted_cryptographer_3.jpg"
		},
		"type": "file"
	},
	{
		"value": "NSA",
		"type": "username"
	},
	{
		"value": "desktop",
		"type": ""
	}],
	"message": "",
	"messageparams": [],
	"link": "https:\/\/localhost:4443\/owncloud\/files\/index.php?dir=%2FContracts%2FGovernments",
	"object_type": "files",
	"object_id": 56142,
	"object_name": "\/Contracts\/Governments\/Most_wanted_cryptographer_3.jpg",
	"subject_prepared": "Shared file \u003Cfile link=\"https:\/\/localhost:4443\/owncloud\/index.php\/apps\/files\/?dir=\/Contracts\/Governments&scrollto=Most_wanted_cryptographer_3.jpg\" id=\"56142\"\u003EContracts\/Governments\/Most_wanted_cryptographer_3.jpg\u003C\/file\u003E was downloaded by \u003Cuser display-name=\"NSA\"\u003ENSA\u003C\/user\u003E via the \u003Cparameter\u003Edesktop\u003C\/parameter\u003E client",
	"message_prepared": "",
	"typeicon": "icon-download",
	"datetime": "2016-03-30T11:08:05+00:00",
	"previews": [{
		"link": "\/owncloud\/index.php\/apps\/files\/?dir=\/Contracts\/Governments&scrollto=Most_wanted_cryptographer_3.jpg",
		"source": "\/owncloud\/index.php\/core\/preview.png?file=\/Contracts\/Governments\/Most_wanted_cryptographer_3.jpg&c=3ed6f28f4d0e2e9c8910e44dd14404bc&y=150&x=150",
		"isMimeTypeIcon": false
	}]
},
```

>An idea would be to add a filter in the web portal's activity feed that let's you select from last hour, day, week, month or year. 

Since that touches the Activity app's UI, I'll open an issue there to see if it's something which can be implemented.
I think it should be possible to get at least a new tab showing only the files which have been downloaded.

**Comment by oparoz** [comment]
@derekblankmccoy  - https://github.com/owncloud/activity/issues/496

**Comment by derekblankmccoy** [comment]
Thanks @oparoz I think we're on a good path.

**Comment by oparoz** [comment]
@derekblankmccoy - Well, that request wasn't approved. It's true that if I add a filter allowing a user to list all downloads, then you can just scroll to find the information you need.

**Comment by nickvergessen** [comment]
Btw if you want to query the API, query the official OCS api only. anything else might change at anytime.

**Comment by oparoz** [comment]
@nickvergessen - you mean your API? https://github.com/owncloud/activity/blob/master/docs/endpoint-v2.md
It's versioned, so shouldn't really change for current clients, no?

**Comment by nickvergessen** [comment]
Only OCS api is stable, the current v2 may or may not become a ocs api in the future, but we are still discussing the format of the output with the clients and w3c comparisons

**Comment by oparoz** [comment]
Thanks @nickvergessen . I saw the discussion in the issues list, but didn't think it would affect the v2. And good job on Activity :).

@derekblankmccoy - There is less info in the OCS API, but it's actually an advantage as it's almost human readable...

```
      {
        "id": 1936,
        "subject": "Shared file Contracts/Governments/4G pineapple.jpg was downloaded by BND via the Web client",
        "message": "",
        "file": "/Contracts/Governments/4G pineapple.jpg",
        "link": "https://localhost:4443/files/index.php?dir=%2FContracts%2FGovernments",
        "date": "2016-03-30T15:06:57+00:00"
      },
```

**Comment by oparoz** [comment]
So, afaik, @gig13's requirements are now met by what I've developed so far:

- [x] User A is notified that User B has downloaded the file and via which method - Web, desktop client or mobile
- [x] It would also be desired to have this tracked in an audit trail **like the Activity** or via a reporting mechanism

@derekblankmccoy - What would you and your partners need in order to be able to close this issue? I've looked at adding an export button in Activity, but since the information is lazy loaded, it will almost never produce what people need. Scripts, etc. are possible, but then it requires a privileged access to the ownCloud installation.

**Comment by derekblankmccoy** [comment]
@oparoz The copy function in the client is good, but it needs more control for it to be functional. It copies the data in comma separated format like this:

date and time in a non workable format, file, action in a sentence written format
Fri Mar 1 07:54:20 2016 GMT,docuements/memo.pdf,Shared file Documents/memo.pdf was downloaded by oparoz via the web client

The info is there, but I can't really do anything with it. If it was like this:

date,time,file,type,user,action,client
01/04/2016,07:54:20,documents/memo.pdf,shared,oparoz,downloaded,web

That would be useful because I could filter the data and do something with it. Remember my scenario of the 500 students? There is no practical way of finding who didn't download the invite without this sort of filtering. I mean unless the activity app had fancy filtering, sorting, and querying, but that would be asking too much in my opinion. A simple extract so I could filter and sort myself is all I would need.

Also, it doesn't go back as far enough as the Activity app shows on the web. The web client is annoying though because I have to scroll down, wait and repeat several times, but at least I can go back a whole year. The problem is I can't export that data easily like the copy button on the client.

That's before considering archiving the data after its copied. There is no way to select how long to go back before exporting the data, so there will be data redundancy. If I could say "show me from 01/03/2016 - 31/03/2016" I could export that and save it as "March 2016.xls". However, this issue could be solved if the data was copied with a usable date format, as I mentioned above. It would copy the whole thing, but I would filter 90% of it out to only show me the last month and I would archive that separately.

**Comment by derekblankmccoy** [comment]
I would also respectfully argue that:

> It would also be desired to have this tracked in an audit trail like the Activity or via a reporting mechanism.

Has not been met as no "audit" can be done if the data cannot be worked with. "A reporting mechanism" also implies that a "report" (filtered query) can be run.

However, the added functionality you added is a massive improvement.

**Comment by oparoz** [comment]
>Has not been met as no "audit" can be done if the data cannot be worked with. "A reporting mechanism" also implies that a "report" (filtered query) can be run.

@derekblankmccoy - Well, the way I read that sentence (and understood from @jospoortvliet's statements) is that Activity is the audit trail. Everything is in the DB and can be queried by any data mining tool, but that's for @gig13 to clarify. Reporting was an alternative proposition in the OP.

But my goal is to close this issue as described in the OP, make the modifications useful and meet the sponsor's requirements, so there has to be a way to easily retrieve the information and neither the client's copy function (Activity API needs to be redesigned, weird timestamps) nor the activity stream (lazy loading, no search) are acceptable solutions.

>have the data on the desktop client go back as far as the web client can

I think it will require the new Activity API, but make sure to open an issue in the client repository if it doesn't exist already.

The big problem right now regarding building a useful export functionality is that the OCS API can't search for information. The only thing you can ask for is the row from which to start the query and the number of rows to return.
So basically, you could dump the last 1000 rows at a regular interval and merge the new batch with what you have...
That's not great.

I see 2 ways forward:

1. Wait for the new OCS API to land and use that either in the client or the web GUI or both
1. Design a custom way to retrieve the information, outside of Activity and not using the OCS API

Any other suggestions? Any preference?

**Comment by derekblankmccoy** [comment]
Well the info is there, just not usable.

If the new API is coming soon, I don't mind waiting for that.
If its pie in the sky, we could make something that takes the data and offers it in a usable manner.

> So basically, you could dump the last 1000 rows at a regular interval and merge the new batch with what you have...

I don't mind that..... If i could select a date range and have it spit it out in workable values.

**Comment by oparoz** [comment]
>If the new API is coming soon, I don't mind waiting for that.

It's targeting 9.1, which means that if you wait for 9.1.1 or 9.1.2 before using it in production, you should be able to get what you need in 4-5 months.
https://github.com/owncloud/activity/issues/494

BTW, my changes will also land in the same version and probably won't be backported, so you should be getting everything at the same time. I you wanted something for 9.0, then a different strategy should be picked.

**Comment by derekblankmccoy** [comment]
Ok so If I understand correctly the new API offers usable data. 
Maybe all we need then is an automatic logger that takes that data and dumps it daily into a yymmdd.csv?

I could then merge the dates I need by using the command prompt.

**Comment by oparoz** [comment]
>Ok so If I understand correctly the new API offers usable data. 

Not quite, the data is usable today, but the API provides a way to extract the data without having to run SQL queries yourself, but unless CSV is added as a response format, you'll get either JSON or XML which you'll still need to convert.
What you will get with the new API though is a way to do what you've just described which is to ask for all the activities recorded during a time period via the Web GUI or a remote client.

But honestly, as mentioned earlier, you can already extract all the data today

1. Install PhpMyAdmin on the server
2. Type in a query which selects the month you're interested in
3. Click on the export button to get your CSV

**Comment by bboule** [comment]
Also... If you are someone that uses MS Excel you can actually connect to the database table directly (old versions of excel need Power Query which is a free add-on) new versions of excel allow you to connect directly to the database, and then allow you to pivot the table in an excel sheet...

Just another option!!!

**Comment by oparoz** [comment]
And @derekblankmccoy there are plenty of formats to choose from if you want to download reports to use straight away.

![plenty_of_formats](https://cloud.githubusercontent.com/assets/323220/14210053/9ad32ce0-f827-11e5-8411-18d107e972fd.png)

| Date | Information |
|:---:	|:-------: |
| 2016-03-30 16:10:48 | [{&quot;56063&quot;:&quot;\/Contracts\/Governments\/QC_Android6_rootkit.rar&quot;},&quot;BND&quot;,&quot;Web&quot;] |
| 2016-03-30 16:11:32 | [{&quot;56051&quot;:&quot;\/Contracts\/Governments\/iOS hacking toolbox.jpg&quot;},&quot;BND&quot;,&quot;Web&quot;] |
| 2016-03-30 17:06:57 | [{&quot;56050&quot;:&quot;\/Contracts\/Governments\/4G pineapple.jpg&quot;},&quot;BND&quot;,&quot;Web&quot;] |

**Comment by derekblankmccoy** [comment]
@oparoz, whats that?

**Comment by oparoz** [comment]
@derekblankmccoy - That's what you can export to using phpMyAdmin

**Comment by derekblankmccoy** [comment]
oh right I see,  the action doesn't seem to be there, I assume you queried for shared file downloads?

**Comment by oparoz** [comment]
>oh right I see, the action doesn't seem to be there, I assume you queried for shared file downloads?

Exactly. All the data for the month of March 2016.

**Comment by oparoz** [comment]
@derekblankmccoy - If you have a test instance and want to give all this a try, I can probably send you a patch and the queries.

**Comment by derekblankmccoy** [comment]
Ok, that sounds good. If I can pull a report with the info needed, we can close this!

**Comment by oparoz** [comment]
@derekblankmccoy - OK, let me know when you have an instance updated with the latest commits from master.

**Comment by derekblankmccoy** [comment]
Im setting up a brand new instance, I'll be back when its done

**Comment by derekblankmccoy** [comment]
It looks like phpmyadmin is no go, but we are going to run a custom script.
What is the query?

**Comment by oparoz** [comment]
>What is the query?

Well, you have lots of options, but here are a couple of examples.

### Information between 2 dates

```
SELECT *
FROM `oc_activity`
WHERE `type` = 'remote_download'
AND `affecteduser` = 'admin'
AND (`timestamp` >= UNIX_TIMESTAMP('2016-03-30')
AND `timestamp` <= UNIX_TIMESTAMP('2016-03-31'))
LIMIT 0 , 100
```

### Information for a specific month

```
SELECT *
FROM `oc_activity`
WHERE `type` = 'remote_download'
AND `affecteduser` = 'admin'
AND (YEAR(FROM_UNIXTIME(`timestamp`)) = 2016
AND MONTH(FROM_UNIXTIME(`timestamp`)) = 03)
LIMIT 0 , 100
```

**Comment by oparoz** [comment]
@derekblankmccoy Any specific hurdles in setting up the test instance?

**Comment by derekblankmccoy** [comment]
Im having issues now. Every time I share a folder with a user, they can see the folder, but if I log out, the shared folder disappears from their ownCloud..... oh dear.

**Comment by oparoz** [comment]
Just saw that the feature freeze for 9.1 is tomorrow, the 24th. I doubt the solution can be tested by then, so it will only be delivered in November, for 9.2.

**Comment by derekblankmccoy** [comment]
Oh, it does work, It's just the retreival of the logs I could not do.

**Comment by oparoz** [comment]
>Oh, it does work.

Does that mean you have a test instance on which to test the patches?

**Comment by derekblankmccoy** [comment]
correct

**Comment by oparoz** [comment]
@derekblankmccoy - Events triggered when files are shared have been accidentally removed. Resolution is tracked in #24937. Once that's fixed, I'll send you patches created against the latest master.

**Comment by oparoz** [comment]
@derekblankmccoy - We're back in business and there are still 12 days before the feature freeze, so this could still make it in if reviewed in time. Make sure you pull the latest commits from master and I'll post a patch you can apply to finally test this.

**Comment by oparoz** [comment]
@derekblankmccoy - Here is [the patch](https://cloud.interfasys.ch/index.php/s/tn7fs9C5hjixtUU). Use `-p0`

**Comment by derekblankmccoy** [comment]
Just manged to get it all patched. It seems to work at first glance!

**Comment by oparoz** [comment]
That's good news :). After a few days, you should have some data to extract via the SQL commands I pasted earlier.

**Comment by derekblankmccoy** [comment]
Cool, so when is it being implemented into a proper release?

**Comment by oparoz** [comment]
Probably in 9.2 in November. 
The first step is for you to validate the fact that it works for you as expected, then I can submit the patches to the repositories which need modifications.

**Comment by oparoz** [comment]
@derekblankmccoy - Can I send my changes to the project? Does that meet your requirements? If this gets merged quickly, I can probably send you patches for versions down to 9.0.

**Comment by derekblankmccoy** [comment]
Tested and working.
I couldn't manage to get LDAP working.
@oparoz is it safe to assume it will track LDAP users just as well?

**Comment by derekblankmccoy** [comment]
Just found an issue with this. It doesn't seem to work with shared folders.

**Comment by oparoz** [comment]
Has been implemented in an app:
https://github.com/nextcloud/files_downloadactivity

**Comment by PVince81** [comment]
From what I see the app is not compatible with ownCloud, so we should keep this ticket open for a solution compatible with ownCloud.

**Comment by PVince81** [comment]
now with multiple link shares enabled in 10.0, it could be possible to generate activities using the link names to make it easier to identify through which link the share was accessed.

**Comment by PVince81** [comment]
@pmaier1

**Comment by xoroz** [comment]
hello everyone, 

I am trying to list all downloaded files activity for one day.
I tried searching the apache log file but I am unable to get the file name/IP.
 grep "GET /index.php/s" /var/log/httpd/owncloud.log-20170725  |grep download
or
 grep "ajax/publicpreview" owncloud.log

occ does not seem to provide that either, can someone help me?

We run 9.1 
Does updating to 10 resolves it?

thank you very much

**Comment by jospoortvliet** [comment]
@xoroz you can try https://apps.nextcloud.com/apps/files_downloadactivity as you can modify the compatibility by editing the app info xml file. Then it might work with ownCloud. Or, if you have the proprietary ownCloud enterprise edition (or free Nextcloud), you can use the auditing feature to get this info from the logs.

**Comment by tomneedham** [comment]
Sounds cool, I'd love a similar feature myself. Sadly cannot load the NC app store as it looks like its down. And looks like the code for that app wasn't touched in over a year :/
